alf/networks/preprocessors.py

```diff
@@ -0,0 +1,130 @@
+# Copyright (c) 2020 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Some network input preprocessors."""
+
+import abc
+import gin
+
+import torch
+import torch.nn as nn
+
+import alf
+from alf.tensor_specs import TensorSpec, BoundedTensorSpec
+from alf.nest.utils import get_outer_rank
+
+
+class InputPreprocessor(nn.Module):
+    """A preprocessor applied to a Network's input.
+
+    It's meant to be applied to either individual inputs or individual
+    TensorSpecs. Mainly for the purpose of input preprocessing and making gin
+    files more convenient to configure.
+
+    Example:
+    In your gin file, below will be possible to configure:
+    input1 (img) -> InputPreprocessor1 -> embed1    ----> EncodingNetwork
+    input2 (action) -> InputPreprocessor2 -> embed2   /   (with `NestCombiner`)
+    """
+
+    def __init__(self, input_tensor_spec, name):
+        super().__init__()
+        assert isinstance(input_tensor_spec, TensorSpec)
+        self._input_tensor_spec = input_tensor_spec
+
+    @abc.abstractmethod
+    def _preprocess(self, tensor):
+        """Preprocess a tensor input.
+
+        Args:
+            tensor (Tensor):
+
+        Returns:
+            a preprocessed tensor
+        """
+        pass
+
+    def forward(self, inputs):
+        """Preprocess either a tensor input or a TensorSpec.
+
+        Args:
+            inputs (TensorSpec or Tensor):
+
+        Returns:
+            Tensor or TensorSpec: if `Tensor`, the returned is the preprocessed
+                result; otherwise it's the tensor spec of the result.
+        """
+        if isinstance(inputs, TensorSpec):
+            tensor = inputs.zeros(outer_dims=(1, ))
+        else:
+            tensor = inputs
+        ret = self._preprocess(tensor)
+        if isinstance(inputs, TensorSpec):
+            return TensorSpec.from_tensor(ret, from_dim=1)
+        return ret
+
+
+@gin.configurable
+class EmbeddingPreprocessor(InputPreprocessor):
+    """A preprocessor that converts the input to an embedding vector. This can
+    be used when the input is a discrete scalar, or a continuous vector to be
+    projected to a different dimension (to have the same length with other
+    vectors). Different from an `EncodingNetwork`, the input can be in the
+    original format from the environment.
```

Comment:
Need to comment that this is not suitable for word embedding or discrete input with large number of discrete values. Or use [nn.Embedding](https://pytorch.org/docs/stable/nn.html#torch.nn.Embedding) to embed discrete input